src/main/java/org/kitteh/irc/client/library/implementation/MemorySTSMachine.java

```diff
@@ -0,0 +1,78 @@
+package org.kitteh.irc.client.library.implementation;
+
+import org.kitteh.irc.client.library.feature.sts.STSClientState;
+import org.kitteh.irc.client.library.feature.sts.STSMachine;
+import org.kitteh.irc.client.library.feature.sts.STSPolicy;
+import org.kitteh.irc.client.library.feature.sts.STSStorageManager;
+import org.kitteh.irc.client.library.util.Sanity;
+
+import javax.annotation.Nonnull;
+import java.net.InetSocketAddress;
+
+/**
+ * "Memory" prefix to distinguish implementation class
+ * from interface, because STSMachine isn't IStsMachine.
+ *
+ * This class implements our FSM in an in-memory fashion,
+ * using Java's data structures.
+ */
+public class MemorySTSMachine implements STSMachine {
+
+    private final STSStorageManager manager;
+    private final InternalClient client;
+    private STSClientState state = STSClientState.UNKNOWN;
+    private STSPolicy policy;
+
+    public MemorySTSMachine(@Nonnull STSStorageManager manager, InternalClient client) {
+        this.client = client;
+        this.manager = Sanity.nullCheck(manager, "Cannot have a null STS persistence manager.");
+    }
+
+    @Nonnull
+    @Override
+    public STSClientState getCurrentState() {
+        return this.state;
+    }
+
+    @Override
+    public void setCurrentState(@Nonnull STSClientState newState) {
+        this.state = Sanity.nullCheck(newState, "Need a valid state for the state machine.");
+        this.step();
+    }
+
+    private void step() {
+        switch (this.state) {
+            case UNKNOWN:
+                throw new IllegalStateException("Unknown state can only be used as an initial state!");
+            case STS_POLICY_CACHED:
+            case STS_PRESENT_RECONNECTING:
+                this.client.getConfig().set(Config.SSL, true);
+                InetSocketAddress oldAddress = this.client.getConfig().get(Config.SERVER_ADDRESS);
+                InetSocketAddress newAddress = new InetSocketAddress(oldAddress.getHostName(), Integer.parseInt(this.policy.getOptions().getOrDefault("port", "6697")));
```

Comment:
Perhaps it would be a good idea to extract `"port"` (and `"duration"`) to constants in `STSPolicy`?